### PR TITLE
[Feature] Optimize TP LMHead with All-to-All

### PR DESCRIPTION
**Pull request by SYChen123**
## Motivation

When the model is deployed with pure DP and dp-attention, **--enable-dp-lm-head will suffer from poor gemm efficiency when batchsize of each rank is small**. For example, in DeepseekV4 Pro, we use dp=8 and --enable-dp-lm-head. In such case, the lmhead gemm itself costs **320us** when bs=36 on each dp rank.
image

In small batch, TP LMHead is a better choice. However, current implementation of TP LMHead will introduce allgather communication on vocab_size. In deepseek v4, each rank will send and receive [bs*8, vocab_size // 8 * 7] = **31MB** data, which costs around **150us**. 
image

Therefore, we propose to use **All2All to replace the allgather + scatter communication in TP LMHead**. With All2All, the data volume that each rank sends and receives decreases to: 

- send: [bs * 7, vocab_size // 8]. When bs=36, it's **3.88MB**.
- receive: [bs, vocab_size // 8 * 7]. When bs=36, it's **3.88MB**.

The all2all costs around 46us in profile.
image

Note that the **all2all optimization is tailored for dp-attention with pure DP parallel**. As for dp-tp hybrid parallel, it's hard to support with significant improvement.

## Modifications

1. Use all2all operation in _get_logits after lmhead computation is completed.
2. Only support pure dp. Dp-lm-head, cp, dp-tp hybrid parallel are not supported to be turned on together.

## Accuracy Tests

GSM8k test on deepseek v4 is OK.
image

## Speed Tests and Profiling

**Profile**
We deploy deepseek-v4-pro on B200 with pd disaggregation. Decode node is launched with pure dp parallel and dp attention. Batch size on each dp rank is 36 and the latency of lmhead is show below.

| Setup | gemm (us) | communication (us) | Total duration of LMHead (us) |
| --- | --- | --- | --- |
| dp-lm-head | 320 | \ | 320 |
| tp-lm-head-allgather | 71 | 151 | 278 |
| tp-lm-head-all2all | **71** | **46** | **169** |

Launch and benchmark command 
```
# prefill
GLOO_SOCKET_IFNAME=eth1 NCCL_SOCKET_IFNAME=eth1 SGLANG_SET_CPU_AFFINITY=1  SGLANG_OPT_SWA_SPLIT_LEAF_ON_INSERT=1 SGLANG_OPT_SWA_EVICT_DROP_PAGE_MARGIN=1 SGLANG_OPT_SWA_RELEASE_LEAF_LOCK_AFTER_WINDOW=1 SGLANG_JIT_DEEPGEMM_FAST_WARMUP=1 SGLANG_OPT_USE_ONLINE_COMPRESS=1 SGLANG_EXPERIMENTAL_ONLINE_C128_MTP=1 SGLANG_DEEPEP_NUM_MAX_DISPATCH_TOKENS_PER_RANK=0 SGLANG_OPT_FIX_MEGA_MOE_MEMORY=1 SGLANG_OPT_USE_DEEPGEMM_MEGA_MOE=1 SGLANG_OPT_DEEPGEMM_MEGA_MOE_NUM_MAX_TOKENS_PER_RANK=8320 SGLANG_OPT_DEEPGEMM_MEGA_MOE_USE_FP4_ACTS=1 SGLANG_OPT_DEEPGEMM_MEGA_MOE_USE_MXF4_KIND=1 sglang serve  --trust-remote-code --model-path ./models/deepseek-ai/DeepSeek-V4-Pro/  --tp 8 --ep 8 --moe-a2a-backend megamoe --enable-dsa-prefill-context-parallel --dsa-prefill-cp-mode round-robin-split --chunked-prefill-size 63488 --max-prefill-tokens 63488 --swa-full-tokens-ratio 0.1 --mem-fraction-static 0.78 --tool-call-parser deepseekv4 --reasoning-parser deepseek-v4 --disaggregation-mode prefill --disaggregation-transfer-backend mooncake  --tokenizer-worker-num 8  --host 0.0.0.0  --port 9091 > log 2>&1 & 

# decode
GLOO_SOCKET_IFNAME=eth1 NCCL_SOCKET_IFNAME=eth1 SGLANG_SET_CPU_AFFINITY=1 SGLANG_OPT_USE_ONLINE_COMPRESS=1 SGLANG_DEEPEP_NUM_MAX_DISPATCH_TOKENS_PER_RANK=0 SGLANG_OPT_FIX_MEGA_MOE_MEMORY=1 SGLANG_OPT_USE_DEEPGEMM_MEGA_MOE=1 SGLANG_OPT_DEEPGEMM_MEGA_MOE_NUM_MAX_TOKENS_PER_RANK=8320 SGLANG_OPT_DEEPGEMM_MEGA_MOE_USE_FP4_ACTS=1 SGLANG_OPT_DEEPGEMM_MEGA_MOE_USE_MXF4_KIND=1 sglang serve --trust-remote-code --model-path ./models/deepseek-ai/DeepSeek-V4-Pro/  --tp 8 --dp 8 --enable-dp-attention --moe-a2a-backend megamoe --mem-fraction-static 0.92 --swa-full-tokens-ratio 0.09 --disaggregation-mode decode --disaggregation-transfer-backend mooncake --enable-hisparse --hisparse-config '{\"top_k\": 1024, \"device_buffer_size\": 4096, \"host_to_device_ratio\": 5}' --tokenizer-worker-num 8 --tool-call-parser deepseekv4 --reasoning-parser deepseek-v4 --host 0.0.0.0 --port 9092 > log 2>&1 & 

# benchmark
python3 -m sglang.bench_serving --model ./models/deepseek-ai/DeepSeek-V4-Pro/ --flush-cache --host 0.0.0.0 --port 9090 --backend sglang --dataset-name generated-shared-prefix --dataset-path ./ShareGPT_V3_unfiltered_cleaned_split.json --num-prompts 1024 --gsp-num-groups 1 --gsp-prompts-per-group 1024 --gsp-system-prompt-len 61440 --gsp-question-len 4096 --gsp-output-len 1024 --request-rate 10 --warmup-requests 10
```

**Benchmark result**
The TPOT decreases **36.97ms** (dp-lmhead) => **36.21ms** (tp-lmhead-allgather) => **35.67ms** (all2all). The E2E latency decreases **3082.24ms** compared with dp-lmhead, **1151.63ms** compared with allgather.

image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
4. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
5. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #32011699808](https://github.com/sgl-project/sglang/actions/runs/32011699808)
Latest PR Test (Extra): :white_check_mark: [Run #32093601089](https://github.com/sgl-project/sglang/actions/runs/32093601089)

### CODE REVIEW

**Review by JustinTong0323 — COMMENTED** [review]
The new all-to-all path must satisfy the flattened-input contract used by PyNccl during CUDA graph capture.

File: python/sglang/srt/layers/logits_processor.py [hunk 1]
```diff
@@ -814,6 +843,45 @@ def _gather_attn_tp_logits(self, logits: torch.Tensor) -> torch.Tensor:
...
+
+        tp_size = get_parallel().tp_size
+        local_rows = local_hidden_states.shape[0]
+        return local_rows > 0 and logits.shape[0] == local_rows * tp_size
+
+    def _tp_lm_head_all_to_all(self, logits: torch.Tensor) -> torch.Tensor:
+        """Exchange only the row block owned by each destination DP rank."""
+        logits = logits.contiguous()
+        all_to_all_output = torch.empty_like(logits)
+        get_tp_group().all_to_all_single(all_to_all_output, logits)
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/logits_processor.py]: Flatten both tensors before calling `all_to_all_single`. CUDA graph capture enables `pynccl_comm`, whose implementation computes `chunk_size = input.numel() // world_size` and then applies `narrow(0, ..., chunk_size)`; for these 2-D logits, that chunk size is larger than dimension 0, so graph capture fails before launching the collective. Pass `all_to_all_output.view(-1)` and `logits.view(-1)` as the DCP path does, and add a CUDA graph smoke test because eager mode falls back to `torch.distributed` and does not expose this failure.

Comment (SYChen123) [reply python/sglang/srt/layers/logits_processor.py]: Done. Thanks for your comments. A test named test_tp_lm_head_all_to_all_cuda_graph has been added. 

If remove the .view(-1) operation, the test will fail at cuadgraph capture. 
image

With the .view(-1) modification, the test is passed.
image

**Comment by SYChen123** [comment]
@JustinTong0323 Hi, thanks for your review. I've resolved your comments and add a cudagraph test case. The CI has all passed. Could you please take a look when available? Thanks

**Comment by SYChen123** [comment]
@JustinTong0323 Could you please trigger the ci again? I resolved the conflicts but CI is not triggered. Thanks

**Comment by SYChen123** [comment]
The all2all path is made default when it's decode node and pure dp attention. Please take a final look. Thanks!
cc @JustinTong0323 @Fridge003

**Review by Fridge003 — APPROVED** [review]

**Comment by SYChen123** [comment]
@Fridge003 I've added an All2All warmup before initializing the kv pool to solve the OOM problem. Now the test_disaggregation_dsv4.py should be passed. 

The all2all warmup aims to preload the necessary p2p transport resources in advance so that it'll be counted in available memory calculation. I also found another bug, which is Deepgemm precompilation will always read gpu0 to check available mem. For intra-node PD where decode is deployed on device4～7, it'll read the wrong device. This is also fixed to ensure the ci test to be passed.

image

**Comment by Eviannn** [comment]
@SYChen123 hi, if token nums across each dp were different, all to all can not work, right? I think this situation is common.

**Comment by SYChen123** [comment]
> @SYChen123 hi, if token nums across each dp were different, all to all can not work, right? I think this situation is common.

@Eviannn Hi, the all2all path is tailored for pure DP parallel on model and TP parallel on lm-head. In this case, the tokens will be allgathered before fed into lm-head. So within TP lm-head, the input token to all2all on each rank is the same and token num is global token num.

**Comment by Eviannn** [comment]
> > @SYChen123 hi, if token nums across each dp were different, all to all can not work, right? I think this situation is common.
> 
> @Eviannn Hi, the all2all path is tailored for pure DP parallel on model and TP parallel on lm-head. In this case, the tokens will be allgathered before fed into lm-head. So within TP lm-head, the input token to all2all on each rank is the same and token num is global token num.

within all gather, token nums are same, but maybe scatter is differ?  _scatter_dp_attn_logits method process this situation, i think

**Comment by SYChen123** [comment]
To guarantee the token num is the same on each rank, here we do the runtime check. In case of cudagraph, all the dp ranks will be padded to the same bucket so token num will be the same. In case of eager, it checks the token num equality on each dp at runtime. If not equal, fallback to the original allgather + scatter.
image

BTW, the reason all2all can be applied is that there's redundant communication in the original allgather + scatter pattern. Let's say bs is the summation of the batchsize on each rank and H is the hidden size. 
- The workflow of allgather + scatter is like `[bs // 8, H] (hidden states generated by model) => AG1 => [bs, H] => LMHead.weight => [bs, H // 8] => AG2 => [bs, H] => scatter => [bs // 8, H]`. The expected shape of LMHead output is [bs//8, H] but after AG2, each rank retains full batch. Then we need to scatter.
- The workflow of all2all is like `[bs // 8, H] (hidden states generated by model) => AG1 => [bs, H] => LMHead.weight => [bs, H // 8] => All2All => [bs // 8, H]`. The output of all2all is exactly what is expected.
